repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py

```diff
@@ -1,5 +1,175 @@
+
+    assert scansourcefiles.is_modified(input_file) == expected_output
+
+
+@pytest.mark.parametrize('run_output,expected_output', (
+    ({'stdout': ['rpm']}, 'rpm'),
+    ({'stdout': ['rpm1', 'rpm2']}, 'rpm1'),
+    (CalledProcessError, ''),
+))
+def test_get_rpm_name(monkeypatch, run_output, expected_output):
+    def mocked_run(cmd, *args, **kwargs):
+        assert cmd == ['rpm', '-qf', '--queryformat', r'%{NAME}\n', input_file]
+        if raise_error:
+            raise CalledProcessError("mocked error", cmd, "result")
+        return run_output
+
+    raise_error = run_output and not isinstance(run_output, dict)
+    input_file = '/file'
+
+    monkeypatch.setattr(scansourcefiles, 'run', mocked_run)
+    monkeypatch.setattr(api, 'current_logger', testutils.logger_mocked())
+
+    assert scansourcefiles._get_rpm_name(input_file) == expected_output
+
+    if raise_error:
+        return
+
+    if len(run_output['stdout']) > 1:
+        assert len(api.current_logger.warnmsg) == 1
+        expected_warnmsg = 'The {} file is owned by multiple rpms: rpm1, rpm2.'.format(input_file)
+        assert api.current_logger.warnmsg[0] == expected_warnmsg
+    else:
+        assert not api.current_logger.warnmsg
+
+
+@pytest.mark.parametrize('input_file,rpm_name,exists,modified,expected_output', (
+    ('/file1', 'rpm', False, True,
+     FileInfo(path='/file1', exists=False, rpm_name='rpm', is_modified=True)),
+
+    ('/file2', '', True, False,
+     FileInfo(path='/file2', exists=True, rpm_name='', is_modified=False))
+))
+def test_scan_file(monkeypatch, input_file, rpm_name, modified, exists, expected_output):
+    monkeypatch.setattr(scansourcefiles, 'is_modified', lambda _: modified)
+    monkeypatch.setattr(scansourcefiles, '_get_rpm_name', lambda _: rpm_name)
+    monkeypatch.setattr(os.path, 'exists', lambda _: exists)
+
+    file_info = scansourcefiles.scan_file(input_file)
+
+    assert isinstance(file_info, FileInfo)
+    assert file_info == expected_output
+
+
+@pytest.mark.parametrize('input_files,rpm_name,exists,modified,expected_output', (
+    ([], '', False, False, []),
+
+    (['/file1'], '', False, False,
+     [FileInfo(path='/file1', exists=False, rpm_name='', is_modified=False)]),
+
+    (['/file1', '/file2'], '', False, False,
+     [FileInfo(path='/file1', exists=False, rpm_name='', is_modified=False),
+      FileInfo(path='/file2', exists=False, rpm_name='', is_modified=False)]),
+
+    (['/file_rpm_modified'], 'rpm', False, True,
+     [FileInfo(path='/file_rpm_modified', exists=False, rpm_name='rpm', is_modified=True)]),
+
+    (['/file_rpm_not_modified'], 'rpm', False, False,
+     [FileInfo(path='/file_rpm_not_modified', exists=False, rpm_name='rpm', is_modified=False)]),
+))
+def test_scan_files(monkeypatch, input_files, rpm_name, exists, modified, expected_output):
+    def scan_file_mocked(input_file):
+        return FileInfo(path=input_file, exists=exists, rpm_name=rpm_name, is_modified=modified)
```

Comment:
this is nice example that there is no reason for duplicated input data

Reply:
Yes I see that now, done.